In [74]:
import pandas as pd
import math

In [75]:
df_main = pd.read_json("./experiment_queries_execution_final_methods.json", orient="records")
df = df_main[~df_main["method_id"].str.contains("boost")]
    

In [76]:
# Number of methods
df["method_id"].unique()


array(['single_elastic_bm25_cross_encoder',
       'single_qdrant_sparse_cross_encoder',
       'single_qdrant_dense_cross_encoder',
       'tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder',
       'triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder'],
      dtype=object)

# Overall

In [77]:
df_thin = df[["method_id", "rr", "time"]]
df_thin.head()

method_id   rr      time
0  single_elastic_bm25_cross_encoder  0.0  1.553253
1  single_elastic_bm25_cross_encoder  0.0  1.448239
2  single_elastic_bm25_cross_encoder  0.0  1.500406
3  single_elastic_bm25_cross_encoder  0.0  1.496796
4  single_elastic_bm25_cross_encoder  0.0  1.138176

In [78]:
df_overall =  df_thin.groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False)
print(df_overall.to_string())

                                                                             method_id        rr      time
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder  0.607824  1.299048
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder  0.605195  1.286085
1                                                    single_qdrant_dense_cross_encoder  0.601799  1.252415
2                                                   single_qdrant_sparse_cross_encoder  0.592412  1.199938
0                                                    single_elastic_bm25_cross_encoder  0.427685  1.275561


### Overall only tuples

In [79]:
df_overall_tuple_only = df_overall[df_overall["method_id"].str.contains("tuple")]
df_overall_tuple_only

method_id        rr      time
4  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...  0.605195  1.286085

### Overall only triples

In [80]:
df_overall_triple_only = df_overall[df_overall["method_id"].str.contains("triple")]
print(df_overall_triple_only.to_string())

                                                                             method_id        rr      time
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder  0.607824  1.299048


# Overall per category



In [81]:
df_thin_per_category = df[["method_id", "class_category", "rr", "time"]]
df_thin_per_category

method_id  \
0                     single_elastic_bm25_cross_encoder   
1                     single_elastic_bm25_cross_encoder   
2                     single_elastic_bm25_cross_encoder   
3                     single_elastic_bm25_cross_encoder   
4                     single_elastic_bm25_cross_encoder   
...                                                 ...   
7535  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
7536  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
7537  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
7538  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
7539  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   

                   class_category        rr      time  
0              instances_100_1000  0.000000  1.553253  
1              instances_100_1000  0.000000  1.448239  
2              instances_100_1000  0.000000  1.500406  
3              instances_100_1000  0.000000  1.496796  
4              instances_100_1000  0.000000  1.138176  
...                           ...       ...       ...  
7535  instances_100000_1000000000  1.000000  1.200883  
7536  instances_100000_1000000000  1.000000  1.117492  
7537  instances_100000_1000000000  0.142857  0.984825  
7538  instances_100000_1000000000  0.100000  1.002219  
7539  instances_100000_1000000000  0.500000  0.957934  

[7540 rows x 4 columns]

In [82]:
df_overall_per_category = df_thin_per_category.groupby(["method_id", "class_category"], as_index=False).mean().sort_values(by=["class_category","rr"], ascending=False)
print(df_overall_per_category.to_string())

                                                                              method_id               class_category        rr      time
15                                                    single_qdrant_dense_cross_encoder           instances_100_1000  0.592936  1.206525
31  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder           instances_100_1000  0.586529  1.251328
39                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder           instances_100_1000  0.585499  1.239876
23                                                   single_qdrant_sparse_cross_encoder           instances_100_1000  0.576575  1.155499
7                                                     single_elastic_bm25_cross_encoder           instances_100_1000  0.437118  1.225313
14                                                    single_qdrant_dense_cross_encoder         instances_1000_10000  0.686049  1.278172
30  triple_fusion_qdrant_dense_0.4_x_qdra

### Only instance categories

In [83]:
print(df_overall_per_category[df_overall_per_category["class_category"].str.startswith("instances")].to_string())

                                                                              method_id               class_category        rr      time
15                                                    single_qdrant_dense_cross_encoder           instances_100_1000  0.592936  1.206525
31  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder           instances_100_1000  0.586529  1.251328
39                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder           instances_100_1000  0.585499  1.239876
23                                                   single_qdrant_sparse_cross_encoder           instances_100_1000  0.576575  1.155499
7                                                     single_elastic_bm25_cross_encoder           instances_100_1000  0.437118  1.225313
14                                                    single_qdrant_dense_cross_encoder         instances_1000_10000  0.686049  1.278172
30  triple_fusion_qdrant_dense_0.4_x_qdra

### Only ancestors cetegories

In [84]:
print(df_overall_per_category[df_overall_per_category["class_category"].str.startswith("ancestors")].to_string())

                                                                              method_id            class_category        rr      time
9                                                     single_qdrant_dense_cross_encoder  ancestors_180_1000000000  0.786388  1.367980
33                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder  ancestors_180_1000000000  0.776843  1.366449
25  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder  ancestors_180_1000000000  0.773638  1.401622
17                                                   single_qdrant_sparse_cross_encoder  ancestors_180_1000000000  0.674279  1.281282
1                                                     single_elastic_bm25_cross_encoder  ancestors_180_1000000000  0.557692  1.440856
16                                                   single_qdrant_sparse_cross_encoder         ancestors_150_180  0.663329  1.372055
32                      tuple_fusion_qdrant_dense_0.4_x_qdrant

### Only children

In [85]:
print(df_overall_per_category[df_overall_per_category["class_category"].str.startswith("children")].to_string())

                                                                              method_id              class_category        rr      time
27  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder       children_10000_100000  0.538558  1.299012
35                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder       children_10000_100000  0.536591  1.285416
11                                                    single_qdrant_dense_cross_encoder       children_10000_100000  0.526674  1.301823
19                                                   single_qdrant_sparse_cross_encoder       children_10000_100000  0.500832  1.163467
3                                                     single_elastic_bm25_cross_encoder       children_10000_100000  0.378086  1.278107
10                                                    single_qdrant_dense_cross_encoder  children_100000_1000000000  0.216814  1.243957
26  triple_fusion_qdrant_dense_0.4_x_qdrant_spar

# Overall - only mechanic queries

In [86]:
df_only_mech_queries = df[df["query_id"].str.startswith("mech")]
df_only_mech_queries

method_id  user_id  class_id  \
672                   single_elastic_bm25_cross_encoder        0  20825628   
673                   single_elastic_bm25_cross_encoder        0  20825628   
674                   single_elastic_bm25_cross_encoder        0  20825628   
675                   single_elastic_bm25_cross_encoder      101  20825628   
676                   single_elastic_bm25_cross_encoder      101  20825628   
...                                                 ...      ...       ...   
7535  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
7536  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
7537  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
7538  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
7539  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   

                     query_id               class_category        rr      time  
672                mech_label           instances_100_1000  1.000000  0.931375  
673                 mech_desc           instances_100_1000  0.000000  1.178072  
674           mech_label_desc           instances_100_1000  1.000000  1.221559  
675   mech_label_user_query_0           instances_100_1000  1.000000  1.105048  
676   mech_label_user_query_1           instances_100_1000  0.000000  1.024904  
...                       ...                          ...       ...       ...  
7535  mech_label_user_query_1  instances_100000_1000000000  1.000000  1.200883  
7536  mech_label_user_query_2  instances_100000_1000000000  1.000000  1.117492  
7537  mech_label_user_query_3  instances_100000_1000000000  0.142857  0.984825  
7538  mech_label_user_query_4  instances_100000_1000000000  0.100000  1.002219  
7539  mech_label_user_query_5  instances_100000_1000000000  0.500000  0.957934  

[4180 rows x 7 columns]

In [87]:
print(df_only_mech_queries[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())


                                                                             method_id        rr      time
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder  0.867031  1.318994
1                                                    single_qdrant_dense_cross_encoder  0.866557  1.247781
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder  0.866240  1.297661
2                                                   single_qdrant_sparse_cross_encoder  0.863801  1.209753
0                                                    single_elastic_bm25_cross_encoder  0.664842  1.309115


### Only labels

In [88]:
df_only_mech_queries_labels_only = df_only_mech_queries[df_only_mech_queries["query_id"].isin(["mech_label"])]
df_only_mech_queries_labels_only

method_id  user_id  class_id  \
672                   single_elastic_bm25_cross_encoder        0  20825628   
681                   single_elastic_bm25_cross_encoder        0   2066754   
690                   single_elastic_bm25_cross_encoder        0     34038   
699                   single_elastic_bm25_cross_encoder        0   7725634   
708                   single_elastic_bm25_cross_encoder        0    602884   
...                                                 ...      ...       ...   
7447  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0   8016240   
7456  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0     23442   
7465  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0     30022   
7474  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0    130019   
7483  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0  28942331   

        query_id               class_category   rr      time  
672   mech_label           instances_100_1000  1.0  0.931375  
681   mech_label         instances_1000_10000  1.0  0.700797  
690   mech_label       instances_10000_100000  1.0  0.633159  
699   mech_label  instances_100000_1000000000  1.0  1.009876  
708   mech_label        children_10000_100000  1.0  0.820918  
...          ...                          ...  ...       ...  
7447  mech_label           instances_100_1000  1.0  0.858936  
7456  mech_label  instances_100000_1000000000  1.0  1.037872  
7465  mech_label         instances_1000_10000  0.5  0.856840  
7474  mech_label       instances_10000_100000  1.0  1.082506  
7483  mech_label           instances_100_1000  1.0  1.007698  

[280 rows x 7 columns]

In [89]:
print(df_only_mech_queries_labels_only[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                             method_id        rr      time
1                                                    single_qdrant_dense_cross_encoder  0.932044  1.033160
0                                                    single_elastic_bm25_cross_encoder  0.923498  1.083049
2                                                   single_qdrant_sparse_cross_encoder  0.918353  1.028742
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder  0.918353  1.104281
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder  0.918353  1.103532


### Only descriptions

In [90]:
df_only_mech_queries_desc_only = df_only_mech_queries[df_only_mech_queries["query_id"].isin(["mech_desc"])]
df_only_mech_queries_desc_only

method_id  user_id  class_id  \
673                   single_elastic_bm25_cross_encoder        0  20825628   
682                   single_elastic_bm25_cross_encoder        0   2066754   
691                   single_elastic_bm25_cross_encoder        0     34038   
700                   single_elastic_bm25_cross_encoder        0   7725634   
709                   single_elastic_bm25_cross_encoder        0    602884   
...                                                 ...      ...       ...   
7448  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0   8016240   
7457  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0     23442   
7466  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0     30022   
7475  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0    130019   
7484  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0  28942331   

       query_id               class_category   rr      time  
673   mech_desc           instances_100_1000  0.0  1.178072  
682   mech_desc         instances_1000_10000  1.0  0.781968  
691   mech_desc       instances_10000_100000  0.0  1.324942  
700   mech_desc  instances_100000_1000000000  1.0  1.013083  
709   mech_desc        children_10000_100000  0.0  2.853606  
...         ...                          ...  ...       ...  
7448  mech_desc           instances_100_1000  1.0  1.120390  
7457  mech_desc  instances_100000_1000000000  1.0  1.147912  
7466  mech_desc         instances_1000_10000  1.0  1.250321  
7475  mech_desc       instances_10000_100000  1.0  1.203862  
7484  mech_desc           instances_100_1000  0.5  1.262597  

[270 rows x 7 columns]

In [91]:
print(df_only_mech_queries_desc_only[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                             method_id        rr      time
2                                                   single_qdrant_sparse_cross_encoder  0.925926  1.264068
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder  0.925926  1.389827
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder  0.925926  1.364825
1                                                    single_qdrant_dense_cross_encoder  0.901235  1.295599
0                                                    single_elastic_bm25_cross_encoder  0.379630  1.403203


### Only labels with descriptions

In [92]:
df_only_mech_queries_label_desc_only = df_only_mech_queries[df_only_mech_queries["query_id"].isin(["mech_label_desc"])]
df_only_mech_queries_label_desc_only

method_id  user_id  class_id  \
674                   single_elastic_bm25_cross_encoder        0  20825628   
683                   single_elastic_bm25_cross_encoder        0   2066754   
692                   single_elastic_bm25_cross_encoder        0     34038   
701                   single_elastic_bm25_cross_encoder        0   7725634   
710                   single_elastic_bm25_cross_encoder        0    602884   
...                                                 ...      ...       ...   
7449  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0   8016240   
7458  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0     23442   
7467  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0     30022   
7476  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0    130019   
7485  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...        0  28942331   

             query_id               class_category   rr      time  
674   mech_label_desc           instances_100_1000  1.0  1.221559  
683   mech_label_desc         instances_1000_10000  1.0  0.826320  
692   mech_label_desc       instances_10000_100000  1.0  1.353524  
701   mech_label_desc  instances_100000_1000000000  1.0  1.121989  
710   mech_label_desc        children_10000_100000  1.0  1.077808  
...               ...                          ...  ...       ...  
7449  mech_label_desc           instances_100_1000  1.0  1.133180  
7458  mech_label_desc  instances_100000_1000000000  1.0  1.230938  
7467  mech_label_desc         instances_1000_10000  1.0  1.272234  
7476  mech_label_desc       instances_10000_100000  1.0  1.211322  
7485  mech_label_desc           instances_100_1000  1.0  1.356907  

[270 rows x 7 columns]

In [93]:
print(df_only_mech_queries_label_desc_only[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                             method_id        rr      time
1                                                    single_qdrant_dense_cross_encoder  1.000000  1.321963
2                                                   single_qdrant_sparse_cross_encoder  1.000000  1.286718
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder  1.000000  1.391488
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder  1.000000  1.374956
0                                                    single_elastic_bm25_cross_encoder  0.814815  1.385894


### Only labels with user queries

In [94]:
df_only_mech_queries_label_user_only = df_only_mech_queries[~df_only_mech_queries["query_id"].isin(["mech_desc", "mech_label", "mech_label_desc"])]
df_only_mech_queries_label_user_only

method_id  user_id  class_id  \
675                   single_elastic_bm25_cross_encoder      101  20825628   
676                   single_elastic_bm25_cross_encoder      101  20825628   
677                   single_elastic_bm25_cross_encoder      101  20825628   
678                   single_elastic_bm25_cross_encoder      101  20825628   
679                   single_elastic_bm25_cross_encoder      101  20825628   
...                                                 ...      ...       ...   
7535  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
7536  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
7537  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
7538  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   
7539  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      114     23442   

                     query_id               class_category        rr      time  
675   mech_label_user_query_0           instances_100_1000  1.000000  1.105048  
676   mech_label_user_query_1           instances_100_1000  0.000000  1.024904  
677   mech_label_user_query_2           instances_100_1000  0.000000  1.280706  
678   mech_label_user_query_3           instances_100_1000  0.000000  1.290999  
679   mech_label_user_query_4           instances_100_1000  1.000000  0.816575  
...                       ...                          ...       ...       ...  
7535  mech_label_user_query_1  instances_100000_1000000000  1.000000  1.200883  
7536  mech_label_user_query_2  instances_100000_1000000000  1.000000  1.117492  
7537  mech_label_user_query_3  instances_100000_1000000000  0.142857  0.984825  
7538  mech_label_user_query_4  instances_100000_1000000000  0.100000  1.002219  
7539  mech_label_user_query_5  instances_100000_1000000000  0.500000  0.957934  

[3360 rows x 7 columns]

In [95]:
print(df_only_mech_queries_label_user_only[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                             method_id        rr      time
1                                                    single_qdrant_dense_cross_encoder  0.847590  1.255863
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder  0.847337  1.325369
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder  0.846353  1.302230
2                                                   single_qdrant_sparse_cross_encoder  0.843318  1.214287
0                                                    single_elastic_bm25_cross_encoder  0.654155  1.314223


### Overall per category

In [96]:
df_only_mech_queries_per_category_thin = df_only_mech_queries[["method_id", "class_category", "rr", "time"]]
df_only_mech_queries_per_category_thin

method_id  \
672                   single_elastic_bm25_cross_encoder   
673                   single_elastic_bm25_cross_encoder   
674                   single_elastic_bm25_cross_encoder   
675                   single_elastic_bm25_cross_encoder   
676                   single_elastic_bm25_cross_encoder   
...                                                 ...   
7535  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
7536  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
7537  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
7538  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
7539  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   

                   class_category        rr      time  
672            instances_100_1000  1.000000  0.931375  
673            instances_100_1000  0.000000  1.178072  
674            instances_100_1000  1.000000  1.221559  
675            instances_100_1000  1.000000  1.105048  
676            instances_100_1000  0.000000  1.024904  
...                           ...       ...       ...  
7535  instances_100000_1000000000  1.000000  1.200883  
7536  instances_100000_1000000000  1.000000  1.117492  
7537  instances_100000_1000000000  0.142857  0.984825  
7538  instances_100000_1000000000  0.100000  1.002219  
7539  instances_100000_1000000000  0.500000  0.957934  

[4180 rows x 4 columns]

In [97]:
df_only_mech_queries_overall_per_category = df_only_mech_queries_per_category_thin.groupby(["method_id", "class_category"], as_index=False).mean().sort_values(by=["class_category", "rr"], ascending=False)
print(df_only_mech_queries_overall_per_category.to_string())

                                                                              method_id               class_category        rr      time
15                                                    single_qdrant_dense_cross_encoder           instances_100_1000  0.839364  1.175513
31  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder           instances_100_1000  0.837663  1.262259
39                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder           instances_100_1000  0.837613  1.243505
23                                                   single_qdrant_sparse_cross_encoder           instances_100_1000  0.826193  1.156337
7                                                     single_elastic_bm25_cross_encoder           instances_100_1000  0.687917  1.261608
14                                                    single_qdrant_dense_cross_encoder         instances_1000_10000  0.970037  1.290175
30  triple_fusion_qdrant_dense_0.4_x_qdra

# Overall - only user queries

In [98]:
df_only_user_queries = df[df["query_id"].str.startswith("user")]
df_only_user_queries

method_id  user_id  class_id  \
0                     single_elastic_bm25_cross_encoder        1  20825628   
1                     single_elastic_bm25_cross_encoder        1  20825628   
2                     single_elastic_bm25_cross_encoder        1  20825628   
3                     single_elastic_bm25_cross_encoder        1  20825628   
4                     single_elastic_bm25_cross_encoder        1  20825628   
...                                                 ...      ...       ...   
6699  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...       14     23442   
6700  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...       14     23442   
6701  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...       14     23442   
6702  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...       14     23442   
6703  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...       14     23442   

          query_id               class_category   rr      time  
0     user_query_0           instances_100_1000  0.0  1.553253  
1     user_query_1           instances_100_1000  0.0  1.448239  
2     user_query_2           instances_100_1000  0.0  1.500406  
3     user_query_3           instances_100_1000  0.0  1.496796  
4     user_query_4           instances_100_1000  0.0  1.138176  
...            ...                          ...  ...       ...  
6699  user_query_1  instances_100000_1000000000  0.5  2.202677  
6700  user_query_2  instances_100000_1000000000  0.5  1.467483  
6701  user_query_3  instances_100000_1000000000  0.0  1.081986  
6702  user_query_4  instances_100000_1000000000  0.0  1.041712  
6703  user_query_5  instances_100000_1000000000  0.0  1.115953  

[3360 rows x 7 columns]

In [99]:
print(df_only_user_queries[["method_id", "rr", "time"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                             method_id        rr      time
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder  0.285357  1.274235
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder  0.280442  1.271684
1                                                    single_qdrant_dense_cross_encoder  0.272428  1.258178
2                                                   single_qdrant_sparse_cross_encoder  0.254791  1.187727
0                                                    single_elastic_bm25_cross_encoder  0.132651  1.233819


### Overall per category

In [100]:
df_only_user_queries_per_category_thin = df_only_user_queries[["method_id", "class_category", "rr", "time"]]
df_only_user_queries_per_category_thin

method_id  \
0                     single_elastic_bm25_cross_encoder   
1                     single_elastic_bm25_cross_encoder   
2                     single_elastic_bm25_cross_encoder   
3                     single_elastic_bm25_cross_encoder   
4                     single_elastic_bm25_cross_encoder   
...                                                 ...   
6699  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
6700  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
6701  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
6702  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   
6703  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   

                   class_category   rr      time  
0              instances_100_1000  0.0  1.553253  
1              instances_100_1000  0.0  1.448239  
2              instances_100_1000  0.0  1.500406  
3              instances_100_1000  0.0  1.496796  
4              instances_100_1000  0.0  1.138176  
...                           ...  ...       ...  
6699  instances_100000_1000000000  0.5  2.202677  
6700  instances_100000_1000000000  0.5  1.467483  
6701  instances_100000_1000000000  0.0  1.081986  
6702  instances_100000_1000000000  0.0  1.041712  
6703  instances_100000_1000000000  0.0  1.115953  

[3360 rows x 4 columns]

In [101]:
df_only_user_queries_overall_per_category = df_only_user_queries_per_category_thin.groupby(["method_id", "class_category"], as_index=False).mean().sort_values(by=["class_category", "rr"], ascending=False)
print(df_only_user_queries_overall_per_category.to_string())

                                                                              method_id               class_category        rr      time
15                                                    single_qdrant_dense_cross_encoder           instances_100_1000  0.284902  1.245289
31  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder           instances_100_1000  0.272611  1.237665
39                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder           instances_100_1000  0.270357  1.235341
23                                                   single_qdrant_sparse_cross_encoder           instances_100_1000  0.264553  1.154452
7                                                     single_elastic_bm25_cross_encoder           instances_100_1000  0.123619  1.179945
30  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder         instances_1000_10000  0.338093  1.286989
14                                       

### Choosing the best match per class of a user

- The idea is that if user writes the 6 queries, and at least one query find the result, the user can use it to model the data structure.
- We want to choose the best query for each of the user assigned classes.
- We can do it by grouping by: method, user, and class.
  - Then we select a maximum for the groups from reciprocal rank.
  - Then the results will be used to again group by a method. 

In [102]:
df_only_user_queries_max_per_user_class = df_only_user_queries[["method_id", "user_id", "class_id", "rr"]].groupby(["method_id", "user_id", "class_id"], as_index=False).max()
df_only_user_queries_max_per_user_class

method_id  user_id  class_id   rr
0                    single_elastic_bm25_cross_encoder        1     34038  0.0
1                    single_elastic_bm25_cross_encoder        1    602884  0.0
2                    single_elastic_bm25_cross_encoder        1   2066754  0.0
3                    single_elastic_bm25_cross_encoder        1   4897819  0.0
4                    single_elastic_bm25_cross_encoder        1   7725634  0.0
..                                                 ...      ...       ...  ...
555  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...       14   8016240  0.0
556  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...       14  11483816  1.0
557  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...       14  18593264  1.0
558  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...       14  28942331  0.2
559  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...       14  51591359  1.0

[560 rows x 4 columns]

### Average rr with only best ranks per classes of users


In [103]:
print(df_only_user_queries_max_per_user_class[["method_id", "rr"]].groupby(["method_id"], as_index=False).mean().sort_values(by="rr", ascending=False).to_string())

                                                                             method_id        rr
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder  0.651962
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder  0.647361
1                                                    single_qdrant_dense_cross_encoder  0.636643
2                                                   single_qdrant_sparse_cross_encoder  0.624419
0                                                    single_elastic_bm25_cross_encoder  0.321662


### Number of classes found per method with only best ranks

- In total we have 56 classes and each class is given to two users.
- Each method have a change of scoring 112 found classes.

In [104]:
df_only_user_queries_max_per_user_class_thin = df_only_user_queries_max_per_user_class[["method_id", "rr"]].copy()
# Round up rr meaning the method found the class.
df_only_user_queries_max_per_user_class_thin["rr"] = df_only_user_queries_max_per_user_class_thin["rr"].apply(lambda x: math.ceil(x))
# Group the results by method -> for a method, each class is presented twice (since two users provided the queries).
print(df_only_user_queries_max_per_user_class_thin.groupby(["method_id"], as_index=False).sum().sort_values(by="rr", ascending=False).to_string())


                                                                             method_id   rr
3  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder  100
4                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder  100
2                                                   single_qdrant_sparse_cross_encoder   95
1                                                    single_qdrant_dense_cross_encoder   92
0                                                    single_elastic_bm25_cross_encoder   43


### For each method and a class we get the number of successful search queries

- There is 12 tries for each method, since we got 2 users and 6 queries per a user.


In [105]:
df1_method_class_successes_thin = df_only_user_queries[["method_id", "class_id", "rr"]]
df1_method_class_successes_thin = df1_method_class_successes_thin.copy()
df1_method_class_successes_thin["rr"] = df1_method_class_successes_thin["rr"].apply(lambda x: math.ceil(x))
print(df1_method_class_successes_thin.groupby(["class_id", "method_id"], as_index=False).sum().to_string())

      class_id                                                                            method_id  rr
0          532                                                    single_elastic_bm25_cross_encoder   0
1          532                                                    single_qdrant_dense_cross_encoder   6
2          532                                                   single_qdrant_sparse_cross_encoder   6
3          532  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder   6
4          532                      tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6_cross_encoder   6
5        11303                                                    single_elastic_bm25_cross_encoder   2
6        11303                                                    single_qdrant_dense_cross_encoder   8
7        11303                                                   single_qdrant_sparse_cross_encoder   7
8        11303  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0

- If we sum this per a class
- We have 5 methods, the class appears twice and 6 queries per class.

In [106]:
print(df1_method_class_successes_thin[["class_id", "rr"]].groupby(["class_id"], as_index=False).sum().sort_values(by="rr", ascending=False).to_string())

     class_id  rr
36   11483816  55
12     130019  51
46   23002039  50
48   29654788  46
49   30325611  45
28    2819161  44
39   15052846  43
22    1348305  43
51   51591359  42
25    1478437  41
38   13442814  40
29    3007842  39
16     294414  38
24    1371150  37
41   18593264  35
20    1134686  35
50   35922124  34
1       11303  33
27    2319498  33
55  124030970  33
17     310899  31
47   28942331  31
45   21191270  31
42   19389637  31
3       24856  29
19    1076052  28
33    5783996  28
43   20755546  28
6       34763  27
4       30022  26
31    3621673  25
37   12890393  25
5       34038  24
0         532  24
9       47521  23
8       44539  23
23    1364302  23
52   63125083  22
30    3231690  20
10      79007  19
18     602884  16
53  123523193  15
40   18557321  14
21    1195098  13
2       23442  13
44   20825628  12
7       43229  11
35    8016240   8
14     134556   8
26    2066754   8
34    7725634   7
15     179049   6
11     124714   5
13     133036   3
32    4897

### Classes the best methods did not find from no rerank results

- triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1
- tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6

In [107]:
# Triple fusion triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1

df1_triple_best = df_only_user_queries_max_per_user_class[df_only_user_queries_max_per_user_class["method_id"].str.contains("triple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.5_x_elastic_bm25_0.1_cross_encoder")][["method_id", "class_id", "rr"]].copy()
df1_triple_best["rr"] = df_only_user_queries_max_per_user_class_thin["rr"].apply(lambda x: math.ceil(x))
df1_triple_best[df1_triple_best["rr"] == 0]

method_id   class_id  rr
339  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...    4897819   0
340  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...    7725634   0
344  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...     133036   0
345  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...     134556   0
351  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...  123691918   0
355  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...    4897819   0
365  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...   20755546   0
367  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...  123691918   0
403  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...     124714   0
405  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...    1195098   0
432  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...      23442   0
443  triple_fusion_qdrant_dense_0.4_x_qdrant_sparse...    8016240   0

- The class it did not find at all:
    - 4897819 - role
    - 123691918 - tool

In [108]:
# Tuple fusion tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6

df1_triple_best = df_only_user_queries_max_per_user_class[df_only_user_queries_max_per_user_class["method_id"].str.contains("tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_0.6")][["method_id", "class_id", "rr"]].copy()
df1_triple_best["rr"] = df_only_user_queries_max_per_user_class_thin["rr"].apply(lambda x: math.ceil(x))
df1_triple_best[df1_triple_best["rr"] == 0]

method_id   class_id  rr
451  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...    4897819   0
452  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...    7725634   0
456  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...     133036   0
457  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...     134556   0
463  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...  123691918   0
467  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...    4897819   0
477  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...   20755546   0
479  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...  123691918   0
515  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...     124714   0
517  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...    1195098   0
544  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...      23442   0
555  tuple_fusion_qdrant_dense_0.4_x_qdrant_sparse_...    8016240   0

- The classes it did not find at all
    - 4897819 - role
    - 123691918 - tool

Interestingly, they are the same as before. Looking into the queries, we tried them manually, and it did not find them.
Thinking, maybe it was since the things were too abstract, the users have a big problem with generating the queries to.

# The least found classes

In [109]:
df[["class_id", "class_category", "rr"]].groupby(["class_id", "class_category"], as_index=False).mean().sort_values(by="rr", ascending=False)

class_id               class_category        rr
12     130019       instances_10000_100000  0.867607
39   15052846           instances_100_1000  0.833466
51   51591359       instances_10000_100000  0.786667
24    1371150           instances_100_1000  0.781743
28    2819161           instances_100_1000  0.774389
49   30325611            ancestors_150_180  0.758025
25    1478437         instances_1000_10000  0.751631
45   21191270  instances_100000_1000000000  0.748713
36   11483816         instances_1000_10000  0.742035
43   20755546     ancestors_180_1000000000  0.740741
48   29654788  instances_100000_1000000000  0.731852
27    2319498       instances_10000_100000  0.708559
50   35922124           instances_100_1000  0.708157
46   23002039         instances_1000_10000  0.699623
22    1348305       instances_10000_100000  0.694718
38   13442814  instances_100000_1000000000  0.684860
55  124030970     ancestors_180_1000000000  0.684637
41   18593264  instances_100000_1000000000  0.677618
52   63125083           instances_100_1000  0.675282
20    1134686       instances_10000_100000  0.674504
19    1076052         instances_1000_10000  0.665309
29    3007842         instances_1000_10000  0.664627
17     310899         instances_1000_10000  0.664239
3       24856         instances_1000_10000  0.659268
42   19389637  instances_100000_1000000000  0.645017
16     294414        children_10000_100000  0.639807
23    1364302           instances_100_1000  0.614821
5       34038       instances_10000_100000  0.608642
4       30022         instances_1000_10000  0.606196
47   28942331           instances_100_1000  0.578553
1       11303         instances_1000_10000  0.566667
26    2066754         instances_1000_10000  0.565432
6       34763       instances_10000_100000  0.558603
34    7725634  instances_100000_1000000000  0.547954
30    3231690       instances_10000_100000  0.529788
15     179049       instances_10000_100000  0.527160
40   18557321            ancestors_150_180  0.526861
0         532  instances_100000_1000000000  0.514533
44   20825628           instances_100_1000  0.506567
9       47521  instances_100000_1000000000  0.506173
10      79007  instances_100000_1000000000  0.472716
31    3621673         instances_1000_10000  0.453418
8       44539         instances_1000_10000  0.446597
33    5783996       instances_10000_100000  0.437543
37   12890393           instances_100_1000  0.429620
7       43229  instances_100000_1000000000  0.427725
2       23442  instances_100000_1000000000  0.415899
18     602884        children_10000_100000  0.352490
35    8016240           instances_100_1000  0.331193
21    1195098       instances_10000_100000  0.327888
54  123691918   children_100000_1000000000  0.267730
13     133036           instances_100_1000  0.248155
11     124714       instances_10000_100000  0.246294
14     134556  instances_100000_1000000000  0.203854
53  123523193           instances_100_1000  0.186829
32    4897819   children_100000_1000000000  0.082354